geoh5py/objects/surveys/electromagnetics/base.py

```diff
@@ -0,0 +1,415 @@
+
+import uuid
+from typing import Any
+
+import numpy as np
+
+from geoh5py.data import FloatData
+from geoh5py.groups import PropertyGroup
+from geoh5py.objects import Curve
+from geoh5py.objects.object_type import ObjectType
+
+
+class BaseEMSurvey(Curve):
+    """
+    A base electromagnetics survey object.
+    """
+
+    __METADATA: dict = {"EM Dataset": {}}
+    __INPUT_TYPE = None
+    __UNITS = None
+    _receivers: BaseEMSurvey | None = None
+    _transmitters: BaseEMSurvey | None = None
+
+    def __init__(self, object_type: ObjectType, **kwargs):
+
+        super().__init__(object_type, **kwargs)
+
+    def add_component_data(self, data: dict) -> list[PropertyGroup]:
+        """
+        Add lists of data components to an EM survey. The name of each component is
+        appended to the metadata 'Property groups'.
+
+        Data channels must be provided for every frequency or time
+        in order specified by
+        :attr:`~geoh5py.objects.surveys.electromagnetics.BaseEMSurvey.channels`.
+        The data channels can be supplied as either a list of
+        :obj:`geoh5py.data.float_data.FloatData` entities or :obj:`uuid.UUID`
+
+        .. code-block:: python
+
+            data = {
+                "Component A": [
+                    data_entity_1,
+                    data_entity_2,
+                ],
+                "Component B": [...],
+            },
+
+        or a nested dictionary of arguments defining new Data entities as defined by the
+        :func:`~geoh5py.objects.object_base.ObjectBase.add_data` method.
+
+        .. code-block:: python
+
+            data = {
+                "Component A": {
+                    time_1: {
+                        'values': [v_11, v_12, ...],
+                        "entity_type": entity_type_A,
+                        ...,
+                    },
+                    time_2: {...},
+                    ...,
+                },
+                "Component B": {...},
+            }
+
+        :param data: Dictionary of data components to be added to the survey.
+
+        :return: List of property groups for all components added.
+        """
+        prop_groups = []
+        if self.channels is None or not self.channels:
+            raise AttributeError(
+                "The 'channels' attribute of an EMSurvey class must be set before the "
+                "'add_component_data' method can be used."
+            )
+
+        if not isinstance(data, dict):
+            raise TypeError(
+                "Input data must be nested dictionaries of components and channels."
+            )
+
+        for name, component_block in data.items():
+            if name in [pg.name for pg in self.property_groups]:
+                raise ValueError(
+                    f"PropertyGroup named '{name}' already exists on the survey entity. "
+                    f"Consider using the 'edit_metadata' method with "
+                    "'Property groups' argument instead."
```

Comment:
**question:** does this mean this method is to create a component, and edit_metadat must be used to update an existing component?
If so, should the function be name something like `create_components()`

Reply:
Changed the method name